libcxx/include/__flat_map/flat_map.h

```diff
@@ -0,0 +1,1280 @@
+    explicit __ctor_uses_allocator_tag() = default;
+  };
+  struct __ctor_uses_allocator_empty_tag {
+    explicit __ctor_uses_allocator_empty_tag() = default;
+  };
+  _LIBCPP_HIDE_FROM_ABI void __sort_and_unique() {
+    auto __zv = ranges::views::zip(__containers_.keys, __containers_.values);
+    ranges::stable_sort(__zv, __compare_, [](const auto& __p) -> decltype(auto) { return std::get<0>(__p); });
+    auto __it   = ranges::unique(__zv, __key_equiv(__compare_)).begin();
+    auto __dist = ranges::distance(__zv.begin(), __it);
+    __containers_.keys.erase(__containers_.keys.begin() + __dist, __containers_.keys.end());
+    __containers_.values.erase(__containers_.values.begin() + __dist, __containers_.values.end());
+  }
+
+  template <class _Allocator, class _KeyCont, class _MappedCont, class... _CompArg>
+  _LIBCPP_HIDE_FROM_ABI
+  flat_map(__ctor_uses_allocator_tag,
+           const _Allocator& __alloc,
+           _KeyCont&& __key_cont,
+           _MappedCont&& __mapped_cont,
+           _CompArg&&... __comp)
+      : __containers_{.keys = std::make_obj_using_allocator<key_container_type>(
+                          __alloc, std::forward<_KeyCont>(__key_cont)),
+                      .values = std::make_obj_using_allocator<mapped_container_type>(
+                          __alloc, std::forward<_MappedCont>(__mapped_cont))},
+        __compare_(std::forward<_CompArg>(__comp)...) {}
+
+  template <class _Allocator, class... _CompArg>
+  _LIBCPP_HIDE_FROM_ABI flat_map(__ctor_uses_allocator_empty_tag, const _Allocator& __alloc, _CompArg&&... __comp)
+      : __containers_{.keys   = std::make_obj_using_allocator<key_container_type>(__alloc),
+                      .values = std::make_obj_using_allocator<mapped_container_type>(__alloc)},
+        __compare_(std::forward<_CompArg>(__comp)...) {}
+
+  template <class _Self, class _Kp>
+  _LIBCPP_HIDE_FROM_ABI static auto __find_impl(_Self&& __self, const _Kp& __key) {
+    auto __it   = __self.lower_bound(__key);
+    auto __last = __self.end();
+    if (__it == __last || __self.__compare_(__key, __it->first)) {
+      return __last;
+    }
+    return __it;
+  }
+
+  template <class _Self>
+  _LIBCPP_HIDE_FROM_ABI static auto __equal_range_return_key_iter(_Self&& __self, const _Key& __key) {
+    auto __it   = ranges::lower_bound(__self.__containers_.keys, __key, __self.__compare_);
+    auto __last = __self.__containers_.keys.end();
+    if (__it == __last || __self.__compare_(__key, *__it)) {
+      return std::make_pair(__it, __it);
+    }
+    return std::make_pair(__it, std::next(__it));
+  }
+
+  template <class _Self, class _Kp>
+  _LIBCPP_HIDE_FROM_ABI static auto __equal_range_return_key_iter(_Self&& __self, const _Kp& __key) {
+    // if the comparator gives different results between  Key(__x) < key2  and __x < key2,
+    // the container might have duplicates w.r.t to  _Kp < Key
+    // TODO: this is the case in Author's test case, but do we really want to support it?
```

Comment:
Supporting this seems strange to me. It would be good for the Standard to be crystal clear about this. Is that not specified?

If it's not specified anywhere, I would argue that this case is not defined by the Standard, aka it's UB. We could catch that UB with `_LIBCPP_ASSERT_SEMANTIC_REQUIREMENT` by checking that the `__key` compares consistently with respect to either key type.

Reply:
The more I think about it, the more I think it is nonsensical to search for a key in a flat_map if sorting the flat_map with that key type would yield a differently sorted or uniqued flat_map. For anything to work sensibly, there's clearly a requirement that the custom `_Kp` type agrees with `_Key` w.r.t. the ordering/uniqueness of the elements in the `flat_map`.

Especially given that this increases the complexity of implementing `equal_range` a lot, IMO we shouldn't support that. We should have an assertion instead.

Reply:
removed support